File: Python/Sorting Algorithms/merge.py
def mergeSort(arr):
    if len(arr) > 1:
        m = len(arr)//2
        l = arr[:m]
        r = arr[m:]
        
        mergeSort(l)
        mergeSort(r)
 
        i=j=k=0

        while i<len(l) and j<len(r):
            if l[i]<=r[j]:
                arr[k]=l[i]
                i+=1
            else:
                arr[k]=r[j]
                j+=1
            k+=1
        
        while i<len(l):
            arr[k]=l[i]
            i+=1
            k+=1
 
        while j<len(r):
            arr[k]=r[j]
            j+=1
            k+=1
    return arr


arr = [3,1,5,7,9,8,10,2,23,22]
print(mergeSort(arr))

File: C++/Sorting/merge.cpp
// Program to implement merge sort
// TC - O(nlog(n))
#include<iostream>
using namespace std;

void merge(int arr[], int s, int m, int f){
    int L=m-s+1;
    int R=f-m;
    int r[R];
    int l[L];

    for(int i=0;i<L;i++){
        l[i]=arr[s+i];
    }

    for(int j=0;j<R;j++){
        r[j]=arr[m+1+j];
    }

    int i=0,j=0,k=s;

    while(i<L && j<R){
        if(l[i]<=r[j]){
            arr[k]=l[i];
            i++;
        } else{
            arr[k]=r[j];
            j++;
        }
        k++;
    }

    while(i<L){
        arr[k]=l[i];
        i++;
        k++;
    }

    while(j<R){
        arr[k]=r[j];
        j++;
        k++;
    }

}

void mergeSort(int arr[],int s,int f){
    if (s<f){
        int m=s+(f-s)/2;
        mergeSort(arr,s,m);
        mergeSort(arr,m+1,f);
        merge(arr,s,m,f);   
    }
}

int main(){
    int a[]={7,2,9,15,11,6,3,1};
    int len=sizeof(a)/sizeof(a[0]);
    mergeSort(a,0,len-1);
    for(auto i:a){
        cout<<i<<" ";
    }
    return 0;
}